torch_xla/csrc/runtime/ifrt_computation_client.cc

```diff
@@ -0,0 +1,643 @@
+#include "torch_xla/csrc/runtime/tf_logging.h"
+#include "torch_xla/csrc/runtime/xla_coordinator.h"
+#include "tsl/profiler/lib/traceme.h"
+#include "xla/client/xla_builder.h"
+#include "xla/client/xla_computation.h"
+#include "xla/layout_util.h"
+#include "xla/literal.h"
+#include "xla/pjrt/distributed/distributed.h"
+#include "xla/pjrt/pjrt_api.h"
+#include "xla/pjrt/pjrt_client.h"
+#include "xla/pjrt/pjrt_executable.h"
+#include "xla/python/ifrt/compiler.h"
+#include "xla/python/ifrt/memory.h"
+#include "xla/python/ifrt/sharding.h"
+#include "xla/python/pjrt_ifrt/pjrt_array.h"
+#include "xla/python/pjrt_ifrt/pjrt_client.h"
+#include "xla/python/pjrt_ifrt/xla_sharding.h"
+#include "xla/shape.h"
+
+using xla::internal::XlaBuilderFriend;
+
+namespace torch_xla {
+namespace runtime {
+
+namespace {
+
+static const std::string spmd_device_str = "SPMD:0";
+
+// Initializes a distributed runtime client if dist_service_addr is specified
+std::shared_ptr<xla::DistributedRuntimeClient>
+MaybeInitializeDistributedRuntimeClient(int local_rank,
+                                        std::string dist_service_addr) {
+  std::shared_ptr<xla::DistributedRuntimeClient> client;
+  if (!dist_service_addr.empty()) {
+    xla::DistributedRuntimeClient::Options options;
+    /* TODO(jonbolin): Use global rank for multi-host setup */
+    options.node_id = local_rank;
+    client = xla::GetDistributedRuntimeClient(dist_service_addr, options);
+    XLA_CHECK(client->Connect().ok())
+        << "Failed to initialize distributed runtime client";
+  }
+  return std::move(client);
+}
+
+// Builds a map from the device's global ordinal to its index in the `devices`
+// array.
+std::unordered_map<int, int> build_index_map(
+    const std::vector<std::string>& devices) {
+  std::unordered_map<int, int> device_index;
+  for (int i = 0; i < devices.size(); ++i) {
+    std::vector<std::string> device_spec = absl::StrSplit(devices[i], ':');
+    XLA_CHECK_EQ(device_spec.size(), 2)
+        << "Invalid device specification: " << devices[i];
+    int global_ordinal = std::stoi(device_spec[1]);
+    device_index[global_ordinal] = i;
+  }
+  return device_index;
+}
+
+}  // namespace
+
+std::string IfrtComputationClient::PjRtDeviceToString(
+    xla::PjRtDevice* const device) const {
+  std::string platform =
+      absl::AsciiStrToUpper(device->client()->platform_name());
+  int ordinal = global_ordinals_.at(device->id());
+  std::string str = absl::StrFormat("%s:%d", platform, ordinal);
+  return str;
+}
+
+std::vector<std::string> IfrtComputationClient::PjRtDevicesToString(
+    absl::Span<xla::PjRtDevice* const> devices) const {
+  std::vector<std::string> strs;
+  strs.reserve(devices.size());
+
+  for (auto* device : devices) {
+    strs.push_back(PjRtDeviceToString(device));
+  }
+
+  return strs;
+}
+
+IfrtComputationClient::IfrtComputationClient() {
+  std::string device_type = sys_util::GetEnvString(env::kEnvPjRtDevice, "");
```

Comment:
let's do a grep of `pjrt` in this file and replace them with `ifrt`. Through I am curious did you intend to query the EnvPjrtDevice here?

Reply:
Yeah, see https://github.com/pytorch/xla/pull/5677#discussion_r1425871830